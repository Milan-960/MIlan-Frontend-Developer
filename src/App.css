.App {
  text-align: center;
}

/* this is for header  */

.fixed-header {
  position: sticky;
  top: 0;
  z-index: 1000;
}

/* this is for info  */

.space-info-container {
  display: flex;
  align-items: center;
  justify-content: space-around;
  font-size: 4rem;
  color: white;
  font-weight: 800;
}

.text-container {
  text-align: center;
}

.launch-button {
  margin-top: 20px;
  padding: 10px 20px;
  border: none;
  background-color: #007bff;
  color: white;
  cursor: pointer;
  float: left;
  font-size: 18px;
  border-radius: 6px;
}

.launch-button:hover {
  background-color: #0056b3;
}

.main {
  height: 100vh;
}

.video {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  min-width: 100%;
  min-height: 100%;
  width: auto;
  height: auto;
  z-index: -100;
  object-fit: cover;
}
