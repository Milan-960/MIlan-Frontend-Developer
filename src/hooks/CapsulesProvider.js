import React, { useState, useEffect, useCallback } from "react";
import { fetchCapsules } from "../services/spacexService";

export const CapsulesContext = React.createContext();

export const CapsulesProvider = ({ children }) => {
  const [allCapsules, setAllCapsules] = useState([]);
  const [capsules, setCapsules] = useState([]);
  const [loading, setLoading] = useState(true);
  const [types, setTypes] = useState([]);
  const [statuses, setStatuses] = useState([]);
  const [launches, setLaunches] = useState([]);

  // Filter capsules based on filter selections
  const handleFiltersChange = useCallback(
    (typeFilter, statusFilter, launchFilter) => {
      let filteredCapsules = allCapsules;

      if (typeFilter) {
        filteredCapsules = filteredCapsules.filter(
          (capsule) => capsule.type && capsule.type === typeFilter.value
        );
      }

      if (statusFilter) {
        filteredCapsules = filteredCapsules.filter(
          (capsule) => capsule.status && capsule.status === statusFilter.value
        );
      }

      if (launchFilter) {
        filteredCapsules = filteredCapsules.filter(
          (capsule) =>
            capsule.original_launch &&
            capsule.original_launch === launchFilter.value
        );
      }

      setCapsules(filteredCapsules);
    },
    [allCapsules] // Only re-calculate the function when allCapsules data changes
  );

  // Fetch capsules data on first render
  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      const data = await fetchCapsules();
      setAllCapsules(data);
      setTypes([...new Set(data.map((item) => item.type))]);
      setStatuses([...new Set(data.map((item) => item.status))]);
      setLaunches([...new Set(data.map((item) => item.original_launch))]);
      setLoading(false);
    };
    fetchData();
  }, []);

  // Set filtered capsules whenever all capsules data is updated
  useEffect(() => {
    setCapsules(allCapsules);
  }, [allCapsules]);

  return (
    <CapsulesContext.Provider
      value={{
        capsules,
        loading,
        types,
        statuses,
        launches,
        handleFiltersChange,
      }}
    >
      {children}
    </CapsulesContext.Provider>
  );
};
