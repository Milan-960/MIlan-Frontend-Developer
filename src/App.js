import React, { useState, useContext } from "react";

import { CapsulesContext } from "./hooks/CapsulesProvider";

import Header from "./components/Header";
import SearchForm from "./components/SearchForm";
import DataGrid from "./components/DataGrid";
import Pagination from "./ui/Pagination";
import Modal from "./ui/Modal";
import SpaceInfo from "./components/Space-Info";

import "./App.css";

const App = () => {
  const { capsules, loading, types, statuses, launches, handleFiltersChange } =
    useContext(CapsulesContext);

  const [selectedCapsule, setSelectedCapsule] = useState(null);
  const [currentPage, setCurrentPage] = useState(1);
  const [itemsPerPage] = useState(10);
  const [isModalOpen, setIsModalOpen] = useState(false);

  // Calculate current items for pagination
  const currentItems = capsules.slice(
    (currentPage - 1) * itemsPerPage,
    currentPage * itemsPerPage
  );

  // Change page number
  const onPageChange = (pageNumber) => setCurrentPage(pageNumber);

  // Handle item selection and open modal
  const handleSelect = (item) => {
    setSelectedCapsule(item);
    setIsModalOpen(true);
  };

  // Close modal and clear selected capsule
  const handleCloseModal = () => {
    setIsModalOpen(false);
    setSelectedCapsule(null);
  };

  return (
    <div className="App">
      <Header />

      <SpaceInfo />

      {loading ? (
        <h1 className="text-center">Loading...</h1>
      ) : (
        <>
          <SearchForm
            onFiltersChange={handleFiltersChange}
            types={types}
            statuses={statuses}
            launches={launches}
          />

          {capsules.length > 0 ? (
            <>
              <DataGrid data={currentItems} onSelect={handleSelect} />
              <Pagination
                currentPage={currentPage}
                itemsPerPage={itemsPerPage}
                totalItems={capsules.length}
                onPageChange={onPageChange}
              />
            </>
          ) : (
            <h1 className="text-center p-4 text-lg font-bold">
              No capsules found
            </h1>
          )}

          {isModalOpen && selectedCapsule && (
            <Modal item={selectedCapsule} onClose={handleCloseModal} />
          )}
        </>
      )}
    </div>
  );
};

export default App;
