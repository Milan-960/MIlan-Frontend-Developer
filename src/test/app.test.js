import React from "react";
import { render } from "@testing-library/react";
import { CapsulesProvider } from "../hooks/CapsulesProvider";
import App from "../App";

jest.mock("../services/spacexService", () => ({
  fetchCapsules: () => Promise.resolve([]),
}));

test("App matches snapshot", async () => {
  const { asFragment } = render(
    <CapsulesProvider>
      <App />
    </CapsulesProvider>
  );

  // Wait for any asynchronous actions (like useEffect) to complete
  await new Promise((resolve) => setTimeout(resolve, 0));

  expect(asFragment()).toMatchSnapshot();
});
