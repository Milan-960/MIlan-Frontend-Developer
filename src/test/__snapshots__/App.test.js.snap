// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="p-5 bg-blue-500 text-white text-center text-xl fixed-header"
    >
      <a
        href="/"
      >
        <img
          alt="spaceX"
          class="w-64"
          src="Space.png"
        />
      </a>
    </header>
    <div
      class="space-info-container"
    >
      <div
        class="text-container"
      >
        <h1>
          UPCOMING LAUNCH 
        </h1>
        <h2>
          STARLINK MISSION
        </h2>
        <button
          class="launch-button"
        >
          Launch
        </button>
      </div>
      <div
        class="main"
      >
        <div>
          
              
          <video
            autoplay=""
            class="video"
            loop=""
            muted=""
            playsinline=""
            preload="metadata"
          >
            
              
            <source
              src="Starship.mp4"
              type="video/mp4"
            />
            
              
          </video>
        </div>
      </div>
    </div>
    <h1
      class="text-center"
    >
      Loading...
    </h1>
  </div>
</DocumentFragment>
`;
